rstudio/rstudio
# Implement Quarto chunk options UI

### Intent

Addresses #11745

### Approach

Chunk / cell options specified using YAML syntax may now be edited via the gear icon / popup UI in both Quarto and R Markdown documents.

<img width="965" alt="editing-ui" src="https://github.com/user-attachments/assets/f85e8be7-e2b6-4c2c-be84-4d4e0648c0c4">

The UI will rewrite options back to the location they were found (e.g. first line or in the YAML block), but new options will go to the YAML location by default for Quarto, and the first line for R Markdown (preserving existing behaviour).

If an option exists both in the first line and in the YAML, then only the YAML one will be written back.

Options are sorted based on existing logic, only tweak I made was to have YAML "label:" be first, if present.

Option names will have periods as separators when in the first line, and dashes in the YAML (fig.cap vs fig-cap). Booleans are uppercase (TRUE/FALSE) in the R code (first line) and lowercase (true/false) in the YAML.

The current implementation does not support the `!expr` technique for including R code. If this or any other technique that the YAML parser doesn't understand are detected when clicking the gear icon, a message will be shown saying that the UI doesn't support the options and to edit manually.

In the future, we can look at customizing the schema used by the YAML parser to support the custom !expr tag (there are examples in the Quarto codebase), but I wanted to get this in before the Sun explodes.

### Automated Tests

There are some, I plan to author more.

### QA Notes

Verify you can edit chunk options whether they are in the first line, or in the YAML lines. Test in both Source mode and Visual mode, in both Quarto and R Markdown (R Markdown has always supported this syntax, not just Quarto).

If you put in a code expression like this (valid for Quarto) the UI should not appear and instead you'll get the warning dialog.

```YAML
#| fig-cap: !expr 'paste("Air", "Quality")' 
```
<img width="767" alt="error-dialog" src="https://github.com/user-attachments/assets/911deb6a-438f-4712-b512-28bea3da7d4f">

Any other text in the `#|` lines that isn't parsable as YAML will also trigger this.

### Documentation

None

### Checklist
